frontend/src/components/Login/Login.jsx

```diff
@@ -1,26 +1,34 @@
-import React from 'react';
+import React, { useEffect, useContext } from 'react';
 import { useHistory } from 'react-router-dom';
 import api from '../../api';
+import { StoreContext } from '../../store';
 
 const Login = () => {
   const history = useHistory();
+  const [store, dispatch] = useContext(StoreContext);
 
   const tryLogin = async e => {
     // Form validation is handled by html5
     e.preventDefault();
     try {
-      await api.post('/login', {
+      const newStore = await api.post('/login', {
         email: e.target.email.value,
         password: e.target.password.value,
       });
-      history.push('/dashboard');
+      dispatch(newStore);
```

Comment:
i think we can call `history.push('/dashboard');` right after dispatch.

Reply:
(Responding as if this will be changed to a `setSession`) -- This is something I've run into before, the issue is that the React Hooks version of `setState` (and I assumed here, `dispatch`) is asynchronous but [can't be made](https://stackoverflow.com/questions/56247433/how-to-use-setstate-callback-on-react-hooks) to return a promise and offers no callback, meaning the `history.push('/dashboard')` can execute before the `state` is set. This becomes a problem when I add in the `<Authenticated>` wrapper for the `/dashboard` route, since `<Authenticated>` will rely on the global state to determine whether it should push the user back to the login screen. So what ultimately ends up happening is that the user is pushed immediately to `/dashboard`, which then sees the `state` is still empty and pushes them back to login. Then `setState` finishes asynchronously but the interface is in the incorrect state.

Shoddy design by the React team imo but long and short is that I have to use `useEffect` to get the behavior I want.

Reply:
Hm...looks a bit strange but as an option you can always use promises since each `then` is asynchronous :-) 

```ts
  api.post('/login', {
    email: e.target.email.value,
    password: e.target.password.value,
  }).then(session => {
    setSession(session);
  })
  .then(() => {
    history.push('/dashboard');
  })
  .catch(error => {
    console.error(error);
  })
```

Reply:
(handling it with useEffect is also fine, so I am resolving this one)

Reply:
I think this would have the same issue (see comments in code below)
```javascript
api.post('/login', {
    email: e.target.email.value,
    password: e.target.password.value,
  }).then(session => {
    setSession(session); // Will return immediately
  })
  .then(() => {
    history.push('/dashboard'); // Will execute before setSession(session) above finishes, resulting in the same problem
  })
  .catch(error => {
    console.error(error);
  })
```